pkg/xds/server/callbacks/dataplane_status_tracker.go

```diff
@@ -114,15 +115,36 @@ func (c *dataplaneStatusTracker) OnStreamRequest(streamID int64, req util_xds.Di
 	state.mu.Lock() // write access to the per Dataplane info
 	defer state.mu.Unlock()
 
-	// infer Dataplane id
 	if state.dataplaneId == (core_model.ResourceKey{}) {
+		var dpType core_model.ResourceType
+		md := core_xds.DataplaneMetadataFromXdsMetadata(req.Metadata())
+
+		// Kuma versions earlier than 1.2.2 didn't always serialize
+		// the DP resource into the Envoy node metadata, so it will
+		// only be available in later kuma-dp versions. In that case,
+		// we can derive the resource type from the proxy type, which
+		// is always present.
+		if md.Resource != nil {
```

Comment:
I'm slightly concerned about putting Dataplane resource to xds metadata every time. For example, we have `DataplaneLifecycle` callback that creates resource in store if it present in metadata: `pkg/xds/server/callbacks/dataplane_lifecycle.go:107`, does it mean we'll override it in the store, for example in Kubernetes?

Why can't we use `md.GetProxyType()` every time?

Reply:
Yeh, I think you are right about this. I liked the idea of the bootstrap always being consistent, but it's not really worth the risk of making the change. I reverted this and always switch on `GetProxyType`